CrossOver 21.0.0

dasmy opened the pull request:
Should build properly. No runtime tests done yet.

dasmy committed:
add version 21.0.0 (untested so far)

dasmy committed:
fix SDL2 path for 21.0.0

dasmy committed:
do not disable vulkan in 21.0.0

dasmy committed:
disable kerberos as it leads to a build error

dasmy commented:
Hi @GabLeRoux, @Gcenx,
This seems to build without any obvious problems. Do you have any thoughts or comments?

Btw.:  I did not runtime-test wine yet.

Gcenx commented:
I did build this last night but figured this morning I’d disabled vkd3d so WineD3D wasn’t usable sigh….

Anyway it’s better to wait for an updated source archive to land as this one is missing MoltenVK/vkd3d and some other components, I’ll let you know once the updated archive lands. (I have an open ticket for this already)

Gcenx commented:
@dasmy & @GabLeRoux of note for CX21.0.0 is the lack of support for macOS Monterey.

The one more things for consideration would be with removing prior versions as CX21.0.0 is much more stable overall compared to prior releases.

dasmy commented:
> I did build this last night but figured this morning I’d disabled vkd3d so WineD3D Vulkan wasn’t usable sigh….
> 
> Anyway it’s better to wait for an updated source archive to land as this one is missing MoltenVK/vkd3d and some other components, I’ll let you know once the updated archive lands. (I have an open ticket for this already)

Yep, I was pretty surprised not to see vkd3d in the archive...

> @dasmy & @GabLeRoux of note for CX21.0.0 is the lack of support for macOS Monterey.

The previous versions don't support Monterey either, do they?

> The one more things for consideration would be with removing prior versions as CX21.0.0 is much more stable overall compared to prior releases.

Definitely - building the older versions seems like a waste of resources. Maybe it will make sense to keep the latest 20.0.x in case someone wants to compare compatibility between versions...

dasmy commented:
Funny enough, I just learned that the build failures for 20.x.x seem to result from a too recent mingw-w64: version 9.x.x of mingw-w64 contains the DirectX code pieces from wine 6.0 which breaks the build of DXVK.

Unfortunately, there does not seem to exist a home-brew formula for mingw-w64 8.x.x. :-(

dasmy committed:
Disable building DXVK in 20.x.x

Build failures seem to result from a too recent mingw-w64: version 9.x.x of mingw-w64 contains the DirectX code pieces from wine 6.0 which breaks the build of DXVK.

Unfortunately, there does not seem to exist a home-brew formula for mingw-w64 8.x.x. :-(

Gcenx commented:
The CX20.0.x build fails are due to sincos add the following patch [msvcrt: Add sincos to importlib](https://github.com/wine-mirror/wine/commit/f0131276474997b9d4e593bbf8c5616b879d3bd5.patch)

The alternative to the patch is passing `-fno-builtin-{sin,cos}{,f}` but I don't like passing multiple flags to `CROSSCFLAGS`, I wouldn't pass `-fcommon` for CX21

That patch is also needed for [wine-stable-6.0.1](https://github.com/Gcenx/macports-wine/blob/main/x11/wine-stable/files/patch-msvcrt_Add_sincos_to_importlib.diff), I had requested it get cherry picked but it didn't qualify for getting fast tracked but at the time at least the M1 GPU patches did qualify to be added.

dasmy committed:
add missing sincos() to msvcrt

see https://github.com/GabLeRoux/macos-crossover-cloud-build/pull/20#issuecomment-893488772

dasmy committed:
remove '-fcommon' from CROSSFLAGS for 20.x.x and 21.x.x

see https://github.com/GabLeRoux/macos-crossover-cloud-build/pull/20#issuecomment-893488772

Gcenx commented on .github/workflows/build_monolithic.yml:
```diff
@@ -106,10 +106,17 @@ jobs:
+        if: startsWith(matrix.CROSS_OVER_VERSION, '20')
+        run: |
+          pushd sources/wine
+          patch -p0 < ${{ github.workspace }}/msvcrt-sincos.patch
```
This should be -p1 when using git formatted patches.

dasmy replied:
Damn... :-D

Gcenx commented on build_local.sh:
```diff
@@ -88,6 +88,11 @@ if [[ ${CROSS_OVER_VERSION} == 20.* ]]; then
 
+    echo "Patch msvcrt to export the missing sincos function, see https://github.com/wine-mirror/wine/commit/f0131276474997b9d4e593bbf8c5616b879d3bd5"
+    pushd sources/wine
+    patch -p0 < ${GITHUB_WORKSPACE}/msvcrt-sincos.patch
```
Same as with build_monolithic.yml

Gcenx commented on .github/workflows/build_local.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       fail-fast: false
       matrix:
-        CROSS_OVER_VERSION: [21.0.0] # , 20.0.4, 20.0.2, 20.0.1, 20.0.0, 19.0.2]
+        CROSS_OVER_VERSION: [21.0.0, 20.0.4, 20.0.2, 20.0.1, 20.0.0, 19.0.2]
```
As the idea is to possibly keep other versions around for checking compatibility it’s only really worth building 21.0.0/20.0.4/19.0.2 anything else would just be a waste of resources.

Gcenx commented on .github/workflows/build_monolithic.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
       fail-fast: false
       matrix:
-        CROSS_OVER_VERSION: [21.0.0] # , 20.0.4, 20.0.2, 20.0.1, 20.0.0, 19.0.2]
+        CROSS_OVER_VERSION: [21.0.0, 20.0.4, 20.0.2, 20.0.1, 20.0.0, 19.0.2]
```
As the idea is to possibly keep other versions around for checking compatibility it’s only really worth building 21.0.0/20.0.4/19.0.2 anything else would just be a waste of resources.

Gcenx reviewed (changes requested):
Now might also be a good time to split the llvm/clang compiling into its own workflow that’s ran first and all others depend on it to avoid compiling it x number of times.

See https://github.com/madewokherd/wine-mono/blob/develop/.github/workflows/c-cpp.yml

dasmy commented:
> https://github.com/madewokherd/wine-mono/blob/develop/.github/workflows/c-cpp.yml

As every source package comes with its own version of llvm/clang: Do you mean building llvm/clang once from the newest package and then using it for the older ones as well?

dasmy committed:
remove version that we are pretty sure about to be not needed any more

Gcenx commented:
> > https://github.com/madewokherd/wine-mono/blob/develop/.github/workflows/c-cpp.yml
> 
> As every source package comes with its own version of llvm/clang: Do you mean building llvm/clang once from the newest package and then using it for the older ones as well?

Basiclly yes as the compiler hasn’t changed since it was initially provided just the source archive layout changed. I’ve build CX21 using the original llvm/clang-8 Xcode toolchain compile without issue.

As upstream is working on implementing its own version of 32on64 that doesn’t require a custom compiler might not longer be required for CX22 at the earliest.

Gcenx commented:
@dasmy & @GabLeRoux if you guys want to quickly see the difference between crossover patch bumps just check my mirror https://github.com/Gcenx/winecx/commit/4939a31816c51cc018f749f7b7c8d99c3a9857b2 Aka crossover-wine-20.0.4

Gcenx commented:
I _think_ `-fcommon` is now only required for CX19.0.2, just don't quote me on that as I can't confirm this currently.

dasmy commented:
> I _think_ `-fcommon` is now only required for CX19.0.2, just don't quote me on that as I can't confirm this currently.

I'll give it a try.

dasmy committed:
see if I can get this modularized (for now without the convoluted if-statements

dasmy committed:
switch to manual run for now

Gcenx commented on .github/workflows/build_monolithic.yml:
```diff
@@ -306,7 +306,7 @@ jobs:
 
       - name: Upload Wine
-        uses: actions/upload-artifact@v2
+        uses: ./github/actions/upload-artifact@v2
```
This workaround lightly won’t be needed for published actions only for the custom ones provided by this repository (I’d think)

dasmy replied:
You are perfectly right - this is the result of a stupid mass-replace of mine...

Gcenx replied:
Oh I’ve done this multiple times with winebuild

Gcenx commented on .github/actions/build_wine.yml:
```diff
@@ -0,0 +1,82 @@
+    # https://developer.apple.com/documentation/xcode-release-notes/xcode-12-release-notes    
+    default: "-g -O2 -Wno-implicit-function-declaration -Wno-deprecated-declarations -Wno-format"
+  ldflags:
+    descriptions: 'CFLAGS to be given to clang'
```
Looks like a copy/paste ,the comment should be LDFLAGS

Gcenx commented:
Once everything is done and working here I think it’s best to squash and rebase into a single commit.

Gcenx commented:
I was afraid this would be the case when slipping the steps.

```
Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/github/actions/configure_wine@v1'. Did you forget to run actions/checkout before running your local action?
```

dasmy committed:
Cache and restore llvm/clang to avoid doing the same thing over and over again

Gcenx commented:
@dasmy I’d in understanding this the current built action I’ll be cached and used if available like what MoltenVK does?

If that’s the idea it might be a good idea to instead move compiling the custom llvm/clang to before anything else and just stick to a fixed url for its sources.

I believe this method isn’t a great fit for multiple CX compiles as each image will need to still complete llvm/clang at times, it would make more sense to split of the llvm/clang compiling into its one action that runs before the others and uploads the archive that then gets reused for all subsequent builds. This could lightly also have a cashed image to speed the one time compile of llvm/clang.

Gcenx commented:
Didn’t mean to close ugh dam phone….

dasmy commented:
> @dasmy I’d in understanding this the current built action I’ll be cached and used if available like what MoltenVK does?
> 
> If that’s the idea it might be a good idea to instead move compiling the custom llvm/clang to before anything else and just stick to a fixed url for its sources.
> 
> I believe this method isn’t a great fit for multiple CX compiles as each image will need to still complete llvm/clang at times, it would make more sense to split of the llvm/clang compiling into its one action that runs before the others and uploads the archive that then gets reused for all subsequent builds. This could lightly also have a cashed image to speed the one time compile of llvm/clang.

Yes, this in general the idea. I am currently reordering the workflow to do exactly that: Have an initial step that restores llvm/clang from cache or restores them (the same for all versions). With my bad experience about quality of the GitHub Action documentation (or my missing capability to read it properly), I am just trying tiny steps one after the other... :-D

dasmy commented:
> > @dasmy I’d in understanding this the current built action I’ll be cached and used if available like what MoltenVK does?
> > If that’s the idea it might be a good idea to instead move compiling the custom llvm/clang to before anything else and just stick to a fixed url for its sources.
> > I believe this method isn’t a great fit for multiple CX compiles as each image will need to still complete llvm/clang at times, it would make more sense to split of the llvm/clang compiling into its one action that runs before the others and uploads the archive that then gets reused for all subsequent builds. This could lightly also have a cashed image to speed the one time compile of llvm/clang.
> 
> Yes, this in general the idea. I am currently reordering the workflow to do exactly that: Have an initial step that restores llvm/clang from cache or restores them (the same for all versions). With my bad experience about quality of the GitHub Action documentation (or my missing capability to read it properly), I am just trying tiny steps one after the other... :-D

... still have to find out, which of the following need to go into the llvm/clang step and which are only for wine (or possibly for both...)

```
        run: |
          # build tools
          brew install  cmake            \
                        ninja            \
                        mingw-w64        \

          # build dependencies for wine / crossover
          brew install  freetype         \
                        bison            \
                        krb5             \
                        faudio           \
                        sdl2             \
                        gphoto2          \
                        sane-backends    \
                        gst-plugins-base \
                        mpg123           \
                        little-cms2      \
                        libpng           \
                        molten-vk

          # dependencies for dxvk
          brew install  coreutils \
                        meson     \
                        glslang
```

Gcenx commented:
For llvm/clang cmake & ninja everything else should already be available. However you don’t even need to install cmake as it’s installed into the images by default anyway.

Gcenx commented:
The following are already installed into the GitHub actions images by default;
- cmake
- freetype
- krb5
- libpng

dasmy committed:
fix indentation

Gcenx commented:
@dasmy as your attempting to use the actions locally the @1 won't be required.

Gcenx committed:
build_monolithic.yml: Don't append @v1 to local actions

Locally hosted actions don't require a version

Gcenx commented on .github/workflows/build_monolithic.yml:
```diff
@@ -181,152 +203,113 @@ jobs:
+          winearch: "wine32on64"
+          build-dir: "${{ env.BUILDROOT }}/wine32on64"
+          crossflags: "-g -O2"
+          configure-params: "--enable-win32on64 --with-wine64=${{ env.BUILDROOT }}/wine64 --without-cms --without-gstreamer --without-gphoto --without-sane --without-krb5"
```
Still require the Vulkan disables for CX21, maybe the CX20/21 configure options can be squashed together.

dasmy replied:
This I don't understand. If you look at https://github.com/GabLeRoux/macos-crossover-cloud-build/runs/3257206753 , win32on64 was configured without all the disable-vk-whatever and it went through...

Gcenx replied:
If you check the last run you’ll see;

```
fatal error: too many errors emitted, stopping now [-ferror-limit=]
36781
20 errors generated.
36782
make: *** [dlls/winemac.drv/vulkan.o] Error 1
36783
make: *** Waiting for unfinished jobs....
36784
Error: Process completed with exit code 2.
```

I have a feeling the prior wine32on64 stages winemac.drv Vulkan was somehow getting skipped, this would match my findings with winebuild on my local system.

Gcenx replied:
Checked the configure log of when wine32on64 completed and I now know why it completed

```
configure: libvulkan and libMoltenVK 64-bit development files not found, Vulkan won't be supported.
```

Prior `export LDFLAGS="-L $(brew --prefix molten-vk)/lib"` wasn't getting passed to wine32on64 configure only for wine64 configure, that solves that annoyance. And yes wine32on64 still doesn't have Vulkan trunking we lightly won't see this until upstream 32to64 implementation is completed.

dasmy replied:
Oh, this makes things clearer. So I should have added "--with-whatever" instead of just removing "--without-whatever" to ensure it fails if the "whatever" component is not available...

Gcenx replied:
To override any configure options already set you can just add --without- later on.

For winebuild I have a text file that contains all desired configure options and wine32on64 appends the needed --without- options to save myself headaches, also doing the same for wine (32Bit) to disable Vulkan/vkd3d etc avoiding build errors.

Gcenx commented on .github/actions/configure_wine/action.yml:
```diff
@@ -0,0 +1,76 @@
+        run: |
+          set -x
+
+          export SDL2_CFLAGS="-I$(brew --prefix sdl2)/include -I$(brew --prefix sdl2)/include/SDL2"
```
This workaround isn’t needed for CX21, however it could also be skipped entirely by adapting the upstream patch https://github.com/wine-mirror/wine/commit/e4fbae832c868e9fcf5a91c58255fe3f4ea1cb30.patch

dasmy replied:
I prefer not to depend on the actual crossover-version inside the action, so that differences between the invocations are obvious directly inside `build_monolithic.yml`. Thus, it's either about properly applying that patch or just leaving things as they are here (doesn't hurt much, I think...). - You are seeing me undecided :-D

Gcenx replied:
It doesn’t make much difference just an alternative.

Gcenx commented on .github/actions/make/action.yml:
```diff
@@ -0,0 +1,15 @@
+      - name: Building in ${{ inputs.build-dir }}
+        run: |
+          pushd ${{ inputs.build-dir }}
+          make -j$(sysctl -n hw.ncpu 2>/dev/null)
```
Maybe tweak this to `$(sysctl -n hw.activecpu)`

dasmy committed:
replace deprecated hw.ncpu by hw.activecpu

Gcenx committed:
Squash CX21 & CX20 32on64 configure options

dasmy committed:
squash 2x config steps

dasmy committed:
consistency: uploading only makes sense if we created an archive

Gcenx commented:
@dasmy seems you pushed the same commit it did to squash CX21/20.

dasmy commented:
> @dasmy seems you pushed the same commit it did to squash CX21/20.

Looks like, but yours was in the wine32on64 section, mine was in the wine64 part

Gcenx commented:
> > @dasmy seems you pushed the same commit it did to squash CX21/20.
> 
> Looks like, but yours was in the wine32on64 section, mine was in the wine64 part

Your right I quickly looked and was confused like how did GitHub allow a duplicate commit without a force push… make sense now.

Gcenx committed:
Move something around test

Gcenx commented:
Moved around the brew dependencies so the needed ones are installed for the relevant stage.

The wine patches are now git format even the distversion.h is now a git format patch. (blame upstream projects for using git format patches)

Next is to figure out what to do about DXVK.....

Gcenx commented:
I believe the following items still need to be worked on before squashing down the commits;
- LLVM/Clang into a separate workflow? (Like wine-mono)
- DXVK (figure out how to build or remove)
- Add back missing wine default tools
- possible winetricks compatibility.

dasmy commented on distversion.patch:
```diff
@@ -0,0 +1,16 @@
++
++#define WINDEBUG_WHAT_HAPPENED_MESSAGE "This can be caused by a problem in the program or a deficiency in Wine. You may want to check <a href=\"http://www.codeweavers.com/compatibility/\">http://www.codeweavers.com/compatibility/</a> for tips about running this application."
++
++#define WINDEBUG_USER_SUGGESTION_MESSAGE "If this problem is not present under Windows and has not been reported yet, you can save the detailed information to a file using the \"Save As\" button, then <a href=\"http://www.codeweavers.com/support/tickets/enter/\">file a bug report</a> and attach that file to the report."
```
As the builds we are creating here are nothing official, we might want to avoid having the codeweavers links in here. - I assume they wouldn't be too happy to be bothered with issue reports for something they did not create themselves?

In my version of `distversion.h` I think I changed the wording a bit.

Gcenx replied:
According to the note for this patch CodeWeaver _did_ take issues for builds but this was prior to 32on64 so maybe that needs to be altered again. I missed that change and just grabbed my copy of the patch.

dasmy committed:
factor build step for llvm/clang into a separate action

dasmy committed:
build llvm/clang in a separate job before building the different wine versions

dasmy committed:
try to debug the unpacking failure

Gcenx commented:
According to the previous action log it shows that nothing was really download. So the extract failed as it was ether an empty archive or simply didn’t exist.

dasmy commented:
> According to the previous action log it shows that nothing was really download. So the extract failed as it was ether an empty archive or simply didn’t exist.

Yes, it's a bit strange (as usual :-) ): The LLVM step says:
```
Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (91.3%)
36
Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (97.6%)
37
Total size of all the files uploaded is 266194120 bytes
38
Finished uploading artifact build-tools-cx. Reported size is 266194120 bytes. There were 0 items that failed to upload
39
Artifact build-tools-cx has been successfully uploaded!
```
If the download file name ore anything were wrong, I would have expected to see an error message. Instead, we are getting an (almost?) empty file - see the output of `ls`...

dasmy commented:
> > According to the previous action log it shows that nothing was really download. So the extract failed as it was ether an empty archive or simply didn’t exist.
> 
> Yes, it's a bit strange (as usual :-) ): The LLVM step says:
> 
> ```
> Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (91.3%)
> 36
> Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (97.6%)
> 37
> Total size of all the files uploaded is 266194120 bytes
> 38
> Finished uploading artifact build-tools-cx. Reported size is 266194120 bytes. There were 0 items that failed to upload
> 39
> Artifact build-tools-cx has been successfully uploaded!
> ```
> 
> If the download file name ore anything were wrong, I would have expected to see an error message. Instead, we are getting an (almost?) empty file - see the output of `ls`...

OK, understood - its a directory, not the actual file... *facepalm again*

dasmy committed:
fix tools download path

dasmy commented:
> > > According to the previous action log it shows that nothing was really download. So the extract failed as it was ether an empty archive or simply didn’t exist.
> > 
> > 
> > Yes, it's a bit strange (as usual :-) ): The LLVM step says:
> > ```
> > Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (91.3%)
> > 36
> > Uploading /Users/runner/work/macos-crossover-cloud-build/macos-crossover-cloud-build/install/build-tools-cx.tar.gz (97.6%)
> > 37
> > Total size of all the files uploaded is 266194120 bytes
> > 38
> > Finished uploading artifact build-tools-cx. Reported size is 266194120 bytes. There were 0 items that failed to upload
> > 39
> > Artifact build-tools-cx has been successfully uploaded!
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If the download file name ore anything were wrong, I would have expected to see an error message. Instead, we are getting an (almost?) empty file - see the output of `ls`...
> 
> OK, understood - its a directory, not the actual file... _facepalm again_

Btw.: In principle, we could also simply download the cache again (the first job step took care of it being valid). - I assume, this might be a bit quicker than the upload-download pair. However, the latter is suggested by the docs. Do you have any preference?

dasmy committed:
oops - forgot to remove a local test

Gcenx commented:
Do what’s recommended just keep the cache step within the llvm/clang action

Gcenx commented:
@dasmy you gotten it working!

So next is figuring out what to do about MoltenVK and possible winetricks compatibility etc as the major things are now handled.

dasmy commented:
> @dasmy you gotten it working!
> 
> So next is figuring out what to do about MoltenVK and possible winetricks compatibility etc as the major things are now handled.

Yes, following your list from above:

> DXVK (figure out how to build or remove)
I added some observations in https://github.com/GabLeRoux/macos-crossover-cloud-build/pull/20/commits/7a2dc8dd3687d87584cae6e7791cfc95378ee956 . Very likely tweaking include paths should help here...
> Add back missing wine default tools
What do you mean by this?
> possible winetricks compatibility
Are there any issues to be expected? - I think I used wibwtricks with crossover 20.0.4

Gcenx commented:
> What do you mean by this?

By default Wine ships some wrapper scripts for easily launching things like winecfg.

> Are there any issues to be expected? - I think I used wibwtricks with crossover 20.0.4

Without setting any environment variables winetricks expects `wine` & `wine64` binaries, so making a symlink/bash script or just make a copy of `wine32on64` & `wine32on64-preloader`. As the output is 6.0 from this winetricks will use any needed workaround(s) for wine-6.0 that should be fine.

Gcenx commented:
@dasmy I’m thinking maybe we merge this and deal with MoltenVK/DXVK in another PR.

What’s are your thoughts on this?

dasmy commented:
> @dasmy I’m thinking maybe we merge this and deal with MoltenVK/DXVK in another PR.
> 
> What’s are your thoughts on this?

@Gcenx, I am fine with the branch being merged. 
Can you maybe do this? - I  am currently on vacation with only mobile phone access :-D

We should just make sure, we don't forget about the open issues, maybe by opening issues ;-)
